### ISSUE TITLE
Creating and inserting IfStmt in place of lines of code I get from parsing MethodExpression

### ISSUE THREAD

**Issue by markiantorno** [javaparser/javaparser#2507]
Hey folks, really appreciate all the work you've put into the library. I've been using it for a lot of really simple repetitive tasks, and it's been working out well. I'm having an issue and I was wondering if you could provide some guidance...

So, I have code that looks like this:

```
public static org.hl7.fhir.r4.model.Patient.ContactComponent convertContactComponent(org.hl7.fhir.dstu2.model.Patient.ContactComponent src) throws FHIRException {
        if (src == null || src.isEmpty())
            return null;
        org.hl7.fhir.r4.model.Patient.ContactComponent tgt = new org.hl7.fhir.r4.model.Patient.ContactComponent();
        VersionConvertor_10_40.copyElement(src, tgt);
        for (org.hl7.fhir.dstu2.model.CodeableConcept t : src.getRelationship()) tgt.addRelationship(VersionConvertor_10_40.convertCodeableConcept(t));
        tgt.setName(VersionConvertor_10_40.convertHumanName(src.getName()));
        for (org.hl7.fhir.dstu2.model.ContactPoint t : src.getTelecom()) tgt.addTelecom(VersionConvertor_10_40.convertContactPoint(t));
        tgt.setAddress(VersionConvertor_10_40.convertAddress(src.getAddress()));
        tgt.setGender(VersionConvertor_10_40.convertAdministrativeGender(src.getGender()));
        tgt.setOrganization(VersionConvertor_10_40.convertReference(src.getOrganization()));
        tgt.setPeriod(VersionConvertor_10_40.convertPeriod(src.getPeriod()));
        return tgt;
    }
```

is it possible to search for all lines of code that contain a "src.set_____()" and wrap them in an if statement that checks for a condition? I've tried doing this with a visitor as follows, but it seems like I'm hacking it instead of doing it properly:

```
 List<MethodDeclaration> methods = classOrInterfaceDeclaration.getMethods();
    methods.forEach(md -> {
        md.accept(new ModifierVisitor<Void>() {
          @Override
          public Visitable visit(MethodCallExpr n, Void arg) {
            super.visit(n, arg);
            if (n.getTokenRange().toString().contains("src.get")) {
              MethodCallExpr topLevelMethodExpr = getTopLevelMethodExpr(n);
              IfStmt ifStmt = new IfStmt();
              ifStmt.setCondition(StaticJavaParser.parseExpression("src.has" + n.getChildNodes().get(1).toString().replaceFirst("get", "") + "()"));
              ifStmt.setThenStmt(StaticJavaParser.parseStatement(topLevelMethodExpr.toString()));
              /*
               * I'm getting errors when I try and parse the expression
               * Exception in thread "main" com.github.javaparser.ParseProblemException: (line 1,col 79) Parse error. 
               * Found <EOF>, expected one of  "%=" "&=" "*=" "++" "+=" "--" "-=" "/=" ";" "<<=" "=" ">>=" ">>>=" "^=" "|="
               * 
               * I'm also puzzled at how to populate the if statement and remove the old line of code at the same time.
               */
              
              System.out.println();

            }
            return n;
          }
        }, null);
      });

  private static MethodCallExpr getTopLevelMethodExpr(MethodCallExpr exp) {
    Optional<Node> node = exp.getParentNode();
    if ((node.isPresent()) && (node.get() instanceof MethodCallExpr)) {
        return getTopLevelMethodExpr((MethodCallExpr) node.get());
    }
    return exp;
  }
```

**Comment by edefazio** [comment]
Little quibble but dont you mean `tgt.set(...)` ? 

Is this (approximately) what you want?
```java
public static ContactComponent convertContactComponent(ContactComponent src) throws FHIRException {
    if (src == null || src.isEmpty())
        return null;
    ContactComponent tgt = new ContactComponent();
    VersionConvertor_10_40.copyElement(src, tgt);
    for (CodeableConcept t : src.getRelationship()) tgt.addRelationship(VersionConvertor_10_40.convertCodeableConcept(t));
    if (src.hasName()) {
        tgt.setName(VersionConvertor_10_40.convertHumanName(src.getName()));
    }
    for (ContactPoint t : src.getTelecom()) tgt.addTelecom(VersionConvertor_10_40.convertContactPoint(t));
    if (src.hasAddress()) {
        tgt.setAddress(VersionConvertor_10_40.convertAddress(src.getAddress()));
    }
    if (src.hasGender()) {
        tgt.setGender(VersionConvertor_10_40.convertAdministrativeGender(src.getGender()));
    }
    if (src.hasOrganization()) {
        tgt.setOrganization(VersionConvertor_10_40.convertReference(src.getOrganization()));
    }
    if (src.hasPeriod()) {
        tgt.setPeriod(VersionConvertor_10_40.convertPeriod(src.getPeriod()));
    }
    return tgt;
}
```
if so, I have a little tool that does this kinda thing , fairly easy:

```java
MethodDeclaration md = "...(the source code)";

//build a pattern for the "final" if statement you want to replace" using $variables$
 $stmt<IfStmt> is = $stmt.ifStmt("if(src.has$Name$()){",
                        "    tgt.set$Name$($container$(src.get$Name$()));",
                       "}");
//search for this pattern in the source (i.e. MethodDeclaration) and transpose/replace the $variables$
//from the source $pattern$ into the target $variables$

$.expressionStmt("tgt.set$Name$($container$(src.get$Name$()));").replaceIn(md, is);
```

as for with using raw JavaParser, I think you are on the right track... but (as a personal preference) generally prefer to Lookup and cache the matching statements using node.walk() to build up a list of matches, then iterate over the list to do the transformations. (it's hard to visit match and transform all in one go)

**Comment by matozoid** [comment]
I've got two general recommendations:
- if you want to know how to examine or build the AST but are lost with all these node types, take a look at https://javaparser.org/inspecting-an-ast/
- if you want to construct the AST for a snippet of code, check out the various parse methods. `StaticJavaParser.parseStatement("if(a){b();}else{c();}").asIfStmt();` will give you a ready-to-go if-statement. (I typed that from memory, so might be slightly wrong.) Personally I find walk/find/remove/replace easier to use than the ModifierVisitor, but that thing works fine as well.
- I would not toString() the tokens like that - better check the fields that contain the data you're looking for.
